File: src/services/safeBroadcastEngine.js
/**
 * 🛡️ SAFE ANTI-BAN WHATSAPP BROADCAST ENGINE
 * Solves WhatsApp number blocking for large lead campaigns
 */

export class SafeBroadcastEngine {
  constructor(config = {}) {
    this.minDelaySec = config.minDelaySec || 15;
    this.maxDelaySec = config.maxDelaySec || 40;
    this.batchSize = config.batchSize || 10;
    this.cooldownMin = config.cooldownMin || 3;
    this.enableTypingSim = config.enableTypingSim !== undefined ? config.enableTypingSim : true;
    this.enableSpintaxJitter = config.enableSpintaxJitter !== undefined ? config.enableSpintaxJitter : true;

    this.queue = [];
    this.sentCount = 0;
    this.failedCount = 0;
    this.isRunning = false;
    this.currentLead = null;
    this.statusText = 'IDLE';
  }

  /**
   * Applies Spintax & Variation Jitter to avoid duplicate message signatures
   */
  applyMessageJitter(adMessage, lead) {
    if (!this.enableSpintaxJitter) return adMessage;

    const greetings = ['Vanakkam', 'Hi', 'Hello', 'Greetings'];
    const closings = [
      'Would you be open to a quick 2-minute call?',
      'Can I share a quick 1-minute video demo?',
      'Would you be interested in discussing this?',
      'Let me know if you would like more details!'
    ];

    const randomGreeting = greetings[Math.floor(Math.random() * greetings.length)];
    const randomClosing = closings[Math.floor(Math.random() * closings.length)];
    const ownerName = (lead.ownerName || 'Sir/Maam').split(' ')[0];

    // Swap greeting dynamically
    let variated = adMessage.replace(/^(Vanakkam|Hi|Hello|Greetings)\s+\w+!/i, `${randomGreeting} ${ownerName}!`);

    // Swap signature closing
    variated = variated.replace(/(Would you be open to a quick 2-minute call.*|Can I share a quick 1-minute video demo.*)/i, randomClosing);

    return variated;
  }

  /**
   * Calculates random delay in milliseconds between min & max range
   */
  getRandomDelayMs() {
    const rangeSec = this.maxDelaySec - this.minDelaySec;
    const randomSec = this.minDelaySec + Math.random() * rangeSec;
    return Math.floor(randomSec * 1000);
  }

  /**
   * Simulates Baileys socket dispatch with safe delay & typing state
   */
  async processBatchItem(item, onProgress) {
    this.currentLead = item;
    const delayMs = this.getRandomDelayMs();
    this.statusText = `Simulating Human Typing (Delay: ${(delayMs / 1000).toFixed(1)}s)...`;
    if (onProgress) onProgress(this.getStatus());

    await new Promise(res => setTimeout(res, delayMs));

    this.sentCount++;
    this.statusText = `Dispatched message to ${item.shopName} (${item.phone})`;
    if (onProgress) onProgress(this.getStatus());
  }

  getStatus() {
    return {
      isRunning: this.isRunning,
      sentCount: this.sentCount,
      failedCount: this.failedCount,
      remainingCount: this.queue.length,
      statusText: this.statusText,
      currentLead: this.currentLead,
      safeDelayRange: `${this.minDelaySec}s - ${this.maxDelaySec}s`,
      batchSize: this.batchSize,
      cooldownMin: this.cooldownMin
    };
  }
}

File: phase2/safe_broadcast_engine.py
"""
🛡️ Python Anti-Ban Safe Broadcast Engine (phase2/safe_broadcast_engine.py)
Implements randomized delay jitter, batching pauses, and typing presence to prevent WhatsApp ban.
"""

import time
import random

class SafeBroadcastEngine:
    def __init__(self, min_delay_sec=15, max_delay_sec=40, batch_size=10, cooldown_min=3):
        self.min_delay_sec = min_delay_sec
        self.max_delay_sec = max_delay_sec
        self.batch_size = batch_size
        self.cooldown_min = cooldown_min
        self.sent_count = 0

    def get_random_delay(self):
        return random.uniform(self.min_delay_sec, self.max_delay_sec)

    def apply_spintax_variation(self, message, owner_name):
        greetings = ["Vanakkam", "Hi", "Hello", "Greetings"]
        random_greeting = random.choice(greetings)
        # Apply dynamic variation to prevent signature matching
        return f"{random_greeting} {owner_name}!\n" + message.split('\n', 1)[-1]

    def send_safe_batch(self, leads, send_callback=None):
        for index, lead in enumerate(leads):
            if index > 0 and index % self.batch_size == 0:
                print(f"⏸️ Anti-Ban Batch Cooldown: Pausing for {self.cooldown_min} minutes...")
                time.sleep(self.cooldown_min * 60)

            delay = self.get_random_delay()
            print(f"⏳ Anti-Ban Safety Delay: Sleeping {delay:.1f}s before sending to {lead.get('shop_name')}...")
            time.sleep(delay)

            if send_callback:
                send_callback(lead)
            
            self.sent_count += 1
            print(f"✅ Dispatched to {lead.get('shop_name')} ({self.sent_count}/{len(leads)})")
